src/soter/openssl/soter_rsa_key_3.c

```diff
@@ -0,0 +1,850 @@
+        goto err;
+    }
+
+    if (!EVP_PKEY_get_bn_param(pkey, OSSL_PKEY_PARAM_RSA_E, &rsa_e)) {
+        res = SOTER_FAIL;
+        goto err;
+    }
+
+    if (BN_is_word(rsa_e, RSA_F4)) {
+        *pub_exp = htobe32(RSA_F4);
+    } else if (BN_is_word(rsa_e, RSA_3)) {
+        *pub_exp = htobe32(RSA_3);
+    } else {
+        res = SOTER_INVALID_PARAMETER;
+        goto err;
+    }
+
+    /* Private exponent */
+    res = bignum_to_bytes(rsa_d, curr_bn, rsa_mod_size);
+    if (SOTER_SUCCESS != res) {
+        goto err;
+    }
+    curr_bn += rsa_mod_size;
+
+    if (!EVP_PKEY_get_bn_param(pkey, OSSL_PKEY_PARAM_RSA_FACTOR1, &rsa_p)) {
+        res = SOTER_FAIL;
+        goto err;
+    }
+
+    if (!EVP_PKEY_get_bn_param(pkey, OSSL_PKEY_PARAM_RSA_FACTOR2, &rsa_q)) {
+        res = SOTER_FAIL;
+        goto err;
+    }
+
+    /* p */
+    res = bignum_to_bytes(rsa_p, curr_bn, rsa_mod_size / 2);
+    if (SOTER_SUCCESS != res) {
+        goto err;
+    }
+    curr_bn += rsa_mod_size / 2;
+
+    /* q */
+    res = bignum_to_bytes(rsa_q, curr_bn, rsa_mod_size / 2);
+    if (SOTER_SUCCESS != res) {
+        goto err;
+    }
+    curr_bn += rsa_mod_size / 2;
+
+    if (!EVP_PKEY_get_bn_param(pkey, OSSL_PKEY_PARAM_RSA_EXPONENT1, &rsa_dmp1)) {
+        res = SOTER_FAIL;
+        goto err;
+    }
+
+    if (!EVP_PKEY_get_bn_param(pkey, OSSL_PKEY_PARAM_RSA_EXPONENT2, &rsa_dmq1)) {
+        res = SOTER_FAIL;
+        goto err;
+    }
+
+    if (!EVP_PKEY_get_bn_param(pkey, OSSL_PKEY_PARAM_RSA_COEFFICIENT1, &rsa_iqmp)) {
+        res = SOTER_FAIL;
+        goto err;
+    }
+
+    /* dp */
+    res = bignum_to_bytes(rsa_dmp1, curr_bn, rsa_mod_size / 2);
+    if (SOTER_SUCCESS != res) {
+        goto err;
+    }
+    curr_bn += rsa_mod_size / 2;
+
+    /* dq */
+    res = bignum_to_bytes(rsa_dmq1, curr_bn, rsa_mod_size / 2);
+    if (SOTER_SUCCESS != res) {
+        goto err;
+    }
+    curr_bn += rsa_mod_size / 2;
+
+    /* qp */
+    res = bignum_to_bytes(rsa_iqmp, curr_bn, rsa_mod_size / 2);
+    if (SOTER_SUCCESS != res) {
+        goto err;
+    }
+    curr_bn += rsa_mod_size / 2;
+
+    /* modulus */
+    res = bignum_to_bytes(rsa_n, curr_bn, rsa_mod_size);
+    if (SOTER_SUCCESS != res) {
+        goto err;
+    }
+
+    memcpy(key->tag, rsa_priv_key_tag(rsa_mod_size), SOTER_CONTAINER_TAG_LENGTH);
+    key->size = htobe32(output_length);
+    soter_update_container_checksum(key);
+    *key_length = output_length;
+    res = SOTER_SUCCESS;
+
+err:
+    BN_clear_free(rsa_d);
+    BN_clear_free(rsa_p);
+    BN_clear_free(rsa_q);
+    BN_clear_free(rsa_dmp1);
+    BN_clear_free(rsa_dmq1);
+    BN_clear_free(rsa_iqmp);
+    BN_free(rsa_n);
+    BN_free(rsa_e);
+
+    //	if (SOTER_SUCCESS != res)
+    //	{
+    //		/* Zero output memory to avoid leaking private key information */
+    //		memset(key, 0, *key_length);
+    //	}
+
+    return res;
+}
+
+/* TODO: Maybe, basic container validation should be put in separate functions outside of
```

Comment:
yes, if we use it in several places. as I understand, previously it was used only in one place?

Reply:
This TODO was in original `soter_rsa_key.c` as well. Is it worth touching in _this_ PR?